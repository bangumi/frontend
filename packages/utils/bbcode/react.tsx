import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
