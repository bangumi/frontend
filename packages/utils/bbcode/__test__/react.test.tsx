import { render, renderNode } from '../react';
import type { VNode } from '../types';

describe('react render vnode', () => {
  test('render multi style', () => {
    const vnode: VNode = {
      type: 'span',
      style: {
        'font-size': '18px',
        'line-height': '18px',
      },
      className: 'l',
      children: ['文字\n换行了'],
    };
    expect(renderNode(vnode)).toMatchSnapshot();
  });

  test('render link', () => {
    const vnode = {
      type: 'a',
      props: {
        href: 'http://chii.in/',
        target: '_blank',
        rel: 'nofollow external noopener noreferrer',
      },
      className: ['l', 'test'],
    };
    expect(renderNode(vnode)).toMatchSnapshot();
  });
  test('render boolean props node', () => {
    const vnode = {
      type: 'input',
      props: {
        type: 'text',
        disabled: true,
      },
    };
    expect(renderNode(vnode)).toMatchSnapshot();
  });
  test('render sticker node', () => {
    const id = '38';
    const vnode = {
      type: 'img',
      props: {
        'sticker-id': id,
        smileid: id,
        alt: `(bgm${id})`,
      },
    };
    expect(renderNode(vnode)).toMatchSnapshot();
  });
  test('render nest', () => {
    const sizeNode: VNode = {
      type: 'span',
      style: {
        'font-size': '18pt',
      },
      children: [
        {
          type: 'a',
          props: {
            href: 'http://chii.in/',
          },
          className: ['l'],
        },
      ],
    };
    const vnode: VNode = {
      type: 'span',
      style: {
        color: 'red',
      },
      children: [sizeNode],
    };
    expect(renderNode(sizeNode)).toMatchSnapshot();
    expect(renderNode(vnode)).toMatchSnapshot();
  });
  test('render mask', () => {
    const vnode: VNode = {
      type: 'span',
      style: {
        'background-color': '#555',
        color: '#555',
        border: '1px solid #555',
      },
      className: 'l',
      children: ['文字\n换行'],
    };
    expect(renderNode(vnode)).toMatchSnapshot();
  });
  test('render pre children', () => {
    const vnode: VNode = {
      type: 'pre',
      children: ['文字\n换行'],
    };
    expect(renderNode(vnode)).toMatchSnapshot();
  });
});

describe('html render bbcode string', () => {
  test('render color size', () => {
    const input = '[color=red][size=18]一般[b]粗[/b]\n   [s]删除[/s][/size][/color]';
    expect(render(input)).toMatchSnapshot();
  });
  test('render mask', () => {
    const input = '我是[mask]马赛克文字[/mask]';
    expect(render(input)).toMatchSnapshot();
  });
  test('render color', () => {
    const input = `我是
[color=red]彩[/color][color=green]色[/color][color=blue]的[/color][color=orange]哟[/color]。`;
    expect(render(input)).toMatchSnapshot();
  });
  test('render size', () => {
    const input = '[size=10]不同[/size][size=14]大小的[/size][size=18]文字[/size]效果也可实现。';
    expect(render(input)).toMatchSnapshot();
  });
  test('render url', () => {
    const input = 'Bangumi 番组计划: [url]http://chii.in/[/url]';
    expect(render(input)).toMatchSnapshot();

    const input2 = `带文字说明的网站链接：
[url=http://chii.in]Bangumi 番组计划[/url]`;
    expect(render(input2)).toMatchSnapshot();
  });

  test('render img', () => {
    const input = `存放于其他网络服务器的图片：
[img]http://chii.in/img/ico/bgm88-31.gif[/img]`;
    expect(render(input)).toMatchSnapshot();
  });
  test('render sticker', () => {
    const input = '(bgm38)(bgm23)(=///=)';
    expect(render(input)).toMatchSnapshot();
  });
  test('render quote', () => {
    const input = '[quote]ss[b]加粗[/b][/quote]';
    expect(render(input)).toMatchSnapshot();
  });
  test('render align', () => {
    const input = '[align=right]右[/align][center]中[/center][left]左[/left]';
    expect(render(input)).toMatchSnapshot();
  });
  test('render @user', () => {
    const input = '[user=1]sai[/user][user]a_little[/user]';
    expect(render(input)).toMatchSnapshot();
  });
  test('render code', () => {
    const input = '[code]ss[b]加粗\n换行了[/b](bgm38) [/fafa [code][/code]';
    expect(render(input)).toMatchSnapshot();
  });
  test('should render sticker', () => {
    expect(render('(bgm01)')).toMatchSnapshot();
    expect(render('(bgm38)')).toMatchSnapshot();
    expect(render('(bgm11)')).toMatchSnapshot();
  });
  test('should render br', () => {
    const input = `
    又名：CUP人生茶话会

    1.上班的时候想要回家的。
    2.学校课上打瞌睡的。(bgm38)
    2.学校课上打瞌睡的。(bgm11)
    2.学校课上打瞌睡的。(bgm12)


    `;

    expect(render(input)).toMatchSnapshot();
  });
  test('should produce unique react keys', () => {
    const checkUniqueKeys = (item: Array<string | React.ReactElement>) => {
      const keys = new Set<React.Key>();
      item.forEach((i) => {
        if (typeof i === 'string') {
          return;
        }
        if (i.key !== null) {
          expect(keys.has(i.key)).toBe(false);
          keys.add(i.key);
        }
        const props = i.props as Record<string, any>;
        if (typeof props.children === 'object') {
          checkUniqueKeys(i.props.children);
        }
      });
    };

    const input = '(bgm38)\n(bgm38)\n';
    const result = render(input);
    checkUniqueKeys(result);
  });
});
