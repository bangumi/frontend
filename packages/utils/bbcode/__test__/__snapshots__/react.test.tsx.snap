// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`html render bbcode string > render @user 1`] = `
Array [
  <a
    className="l"
    href="/user/1"
    style={Object {}}
  >
    @sai
  </a>,
  <a
    className="l"
    href="/user/a_little"
    style={Object {}}
  >
    @a_little
  </a>,
]
`;

exports[`html render bbcode string > render align 1`] = `
Array [
  <p
    style={
      Object {
        "textAlign": "right",
      }
    }
  >
    右
  </p>,
  <p
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    中
  </p>,
  <p
    style={
      Object {
        "textAlign": "left",
      }
    }
  >
    左
  </p>,
]
`;

exports[`html render bbcode string > render code 1`] = `
Array [
  <pre
    style={Object {}}
  >
    <React.Fragment>
      ss[b]加粗
      <br />
      换行了[/b](bgm38) [/fafa [code]
    </React.Fragment>
  </pre>,
]
`;

exports[`html render bbcode string > render color 1`] = `
Array [
  <React.Fragment>
    我是
    <br />
    
  </React.Fragment>,
  <span
    style={
      Object {
        "color": "red",
      }
    }
  >
    彩
  </span>,
  <span
    style={
      Object {
        "color": "green",
      }
    }
  >
    色
  </span>,
  <span
    style={
      Object {
        "color": "blue",
      }
    }
  >
    的
  </span>,
  <span
    style={
      Object {
        "color": "orange",
      }
    }
  >
    哟
  </span>,
  "。",
]
`;

exports[`html render bbcode string > render color size 1`] = `
Array [
  <span
    style={
      Object {
        "color": "red",
      }
    }
  >
    <span
      style={
        Object {
          "fontSize": "18px",
          "lineHeight": "18px",
        }
      }
    >
      一般
      <strong
        style={Object {}}
      >
        粗
      </strong>
      <React.Fragment>
        
        <br />
           
      </React.Fragment>
      <span
        style={
          Object {
            "textDecoration": "line-through",
          }
        }
      >
        删除
      </span>
    </span>
  </span>,
]
`;

exports[`html render bbcode string > render img 1`] = `
Array [
  <React.Fragment>
    存放于其他网络服务器的图片：
    <br />
    
  </React.Fragment>,
  <img
    className="code"
    src="http://chii.in/img/ico/bgm88-31.gif"
    style={Object {}}
  />,
]
`;

exports[`html render bbcode string > render mask 1`] = `
Array [
  "我是",
  <span
    className="bgm-mask"
    style={Object {}}
  >
    马赛克文字
  </span>,
]
`;

exports[`html render bbcode string > render quote 1`] = `
Array [
  <div
    className="quote"
    style={Object {}}
  >
    <q
      style={Object {}}
    >
      ss
      <strong
        style={Object {}}
      >
        加粗
      </strong>
    </q>
  </div>,
]
`;

exports[`html render bbcode string > render size 1`] = `
Array [
  <span
    style={
      Object {
        "fontSize": "10px",
        "lineHeight": "10px",
      }
    }
  >
    不同
  </span>,
  <span
    style={
      Object {
        "fontSize": "14px",
        "lineHeight": "14px",
      }
    }
  >
    大小的
  </span>,
  <span
    style={
      Object {
        "fontSize": "18px",
        "lineHeight": "18px",
      }
    }
  >
    文字
  </span>,
  "效果也可实现。",
]
`;

exports[`html render bbcode string > render sticker 1`] = `
Array [
  <img
    alt="(bgm38)"
    smileid="54"
    src="https://lain.bgm.tv/img/smiles/tv/15.gif"
    style={Object {}}
  />,
  <img
    alt="(bgm23)"
    smileid="39"
    src="https://lain.bgm.tv/img/smiles/bgm/23.gif"
    style={Object {}}
  />,
  <img
    alt="(=///=)"
    smileid="13"
    src="https://lain.bgm.tv/img/smiles/13.gif"
    style={Object {}}
  />,
]
`;

exports[`html render bbcode string > render url 1`] = `
Array [
  "Bangumi 番组计划: ",
  <a
    className="bgm-link"
    href="http://chii.in/"
    style={Object {}}
  >
    http://chii.in/
  </a>,
]
`;

exports[`html render bbcode string > render url 2`] = `
Array [
  <React.Fragment>
    带文字说明的网站链接：
    <br />
    
  </React.Fragment>,
  <a
    className="bgm-link"
    href="http://chii.in"
    style={Object {}}
  >
    Bangumi 番组计划
  </a>,
]
`;

exports[`html render bbcode string > should render br 1`] = `
Array [
  <React.Fragment>
    
    <br />
        又名：CUP人生茶话会
    <br />
    
    <br />
        1.上班的时候想要回家的。
    <br />
        2.学校课上打瞌睡的。
  </React.Fragment>,
  <img
    alt="(bgm38)"
    smileid="54"
    src="https://lain.bgm.tv/img/smiles/tv/15.gif"
    style={Object {}}
  />,
  <React.Fragment>
    
    <br />
        2.学校课上打瞌睡的。
  </React.Fragment>,
  <img
    alt="(bgm11)"
    smileid="27"
    src="https://lain.bgm.tv/img/smiles/bgm/11.gif"
    style={Object {}}
  />,
  <React.Fragment>
    
    <br />
        2.学校课上打瞌睡的。
  </React.Fragment>,
  <img
    alt="(bgm12)"
    smileid="28"
    src="https://lain.bgm.tv/img/smiles/bgm/12.png"
    style={Object {}}
  />,
  <React.Fragment>
    
    <br />
    
    <br />
    
    <br />
        
  </React.Fragment>,
]
`;

exports[`html render bbcode string > should render sticker 1`] = `
Array [
  <img
    alt="(bgm01)"
    smileid="17"
    src="https://lain.bgm.tv/img/smiles/bgm/01.png"
    style={Object {}}
  />,
]
`;

exports[`html render bbcode string > should render sticker 2`] = `
Array [
  <img
    alt="(bgm38)"
    smileid="54"
    src="https://lain.bgm.tv/img/smiles/tv/15.gif"
    style={Object {}}
  />,
]
`;

exports[`html render bbcode string > should render sticker 3`] = `
Array [
  <img
    alt="(bgm11)"
    smileid="27"
    src="https://lain.bgm.tv/img/smiles/bgm/11.gif"
    style={Object {}}
  />,
]
`;

exports[`react render vnode > render boolean props node 1`] = `
<input
  disabled={true}
  style={Object {}}
  type="text"
/>
`;

exports[`react render vnode > render link 1`] = `
<a
  className="l test"
  href="http://chii.in/"
  rel="nofollow external noopener noreferrer"
  style={Object {}}
  target="_blank"
/>
`;

exports[`react render vnode > render mask 1`] = `
<span
  className="l"
  style={
    Object {
      "backgroundColor": "#555",
      "border": "1px solid #555",
      "color": "#555",
    }
  }
>
  <React.Fragment>
    文字
    <br />
    换行
  </React.Fragment>
</span>
`;

exports[`react render vnode > render multi style 1`] = `
<span
  className="l"
  style={
    Object {
      "fontSize": "18px",
      "lineHeight": "18px",
    }
  }
>
  <React.Fragment>
    文字
    <br />
    换行了
  </React.Fragment>
</span>
`;

exports[`react render vnode > render nest 1`] = `
<span
  style={
    Object {
      "fontSize": "18pt",
    }
  }
>
  <a
    className="l"
    href="http://chii.in/"
    style={Object {}}
  />
</span>
`;

exports[`react render vnode > render nest 2`] = `
<span
  style={
    Object {
      "color": "red",
    }
  }
>
  <span
    style={
      Object {
        "fontSize": "18pt",
      }
    }
  >
    <a
      className="l"
      href="http://chii.in/"
      style={Object {}}
    />
  </span>
</span>
`;

exports[`react render vnode > render pre children 1`] = `
<pre
  style={Object {}}
>
  <React.Fragment>
    文字
    <br />
    换行
  </React.Fragment>
</pre>
`;

exports[`react render vnode > render sticker node 1`] = `
<img
  alt="(bgm38)"
  smileid="38"
  sticker-id="38"
  style={Object {}}
/>
`;
